Implement Proxmox builder

carlpett opened the pull request:
This pull request implements a builder for [Proxmox VE](https://www.proxmox.com/en/proxmox-ve).

The PR is not yet ready to merge since I have a couple of patches to the Proxmox API library that are not yet merged (so the dependency is not vendored yet), but I'm looking forward to comments on the Packer implementation while that is fixed :)
I also haven't implemented any tests yet (:flushed:), would appreciate pointers on what parts of a new builder should be tested.

hashicorp-cla commented:
[![CLA assistant check](https://cla.hashicorp.com/pull/badge/not_signed)](https://cla.hashicorp.com/hashicorp/packer?pullRequest=7391)


Thank you for your submission! We require that all contributors sign our Contributor License Agreement ("CLA") before we can accept the contribution. [Read and sign the agreement](https://cla.hashicorp.com/hashicorp/packer?pullRequest=7391)

[Learn more about why HashiCorp requires a CLA and what the CLA includes](https://www.hashicorp.com/cla)

<sub>Have you signed the CLA already but the status is still pending? [Recheck it](https://cla.hashicorp.com/check/hashicorp/packer?pullRequest=7391).</sub>

SwampDragons commented:
Thanks for this!  Looks like tests are failing because of a vendoring issue; mind looking into that?

SwampDragons commented:
Code overall looks pretty good. 

As far as unit tests go, we like at a minimum to see good coverage of the validation being done in NewConfig()

Extra credit for making an api mock and testing some of the behavior in the run() portion of your various steps.

I like your trick with StepSuccess. Would you mind getting hold of some credentials so I can build and test once you're ready?

SwampDragons commented:
Oh, we also need docs for this. Take a look at the docs in website/src for examples.

carlpett commented:
Thanks for the first look @SwampDragons! Good to have some validation early that things aren't completely in the wrong direction :)
And thanks for the guidance on what to test. I'll see what I can do apart from the config validation. Most of the heavy lifting is in the library, but there might be a bunch of things that can be done nevertheless.

Good point about the docs, adding that to the list as well :) 

As for credentials, this is on-prem software, and I don't have a public installation unfortunately. It does work decently well in nested virtualization setups though, and is pretty much download and next-next-finish to have a simple working environment.

SwampDragons commented:
okay, if you can point me to some docs so I can try to get it working that'd be good. Barring that, maybe just show me some example log output so I can have some proof of life for the builder 😂

carlpett commented:
Their getting started guide is [here](https://www.proxmox.com/en/proxmox-ve/get-started), let me know if you have any questions or run into any issues!
I've pushed commits with docs now, as well as some simple config validation testing and a thought about how to test steps. Does it look like a good approach to you?

SwampDragons commented:
Yeah I think this is looking really good.

carlpett commented:
Nice! I also think it worked out quite nicely. Added a bunch more tests now (and found a handful of bugs in the process :flushed:). The main untested functionality now is the start_vm step. Unfortunately the API is not very suited to mocking there, with package-level functions and a some ancillary types.
I'll see what I can do about it, but it requires a bit more thought.

SwampDragons commented:
Don't worry too much about mocking. If the majority of the control logic is tested, I'm happy with that.

carlpett commented:
@SwampDragons I rebased on master, updated the `Builder` implementation to keep up with the change in the interface, and added the vendor dir.

The failing provisioner tests are failing on master for me as well, so hopefully not related?

carlpett commented:
Realized I still had this marked as WIP. It is now ready for review!

carlpett commented:
@SwampDragons Anything else needed to progress here?

SwampDragons commented:
Ack sorry, no, I'm just way behind on my reviews.

SwampDragons commented:
Rebased and pushed again to clear tests.

carlpett commented:
No worries, just let me know there's any action expected from me :)

SwampDragons commented:
Hmm, actually maybe yes. After rebasing on master (which is currently not failing tests), tests are still failing in the same way. It looks to me like the build failures are actually specific to your branch, and seem to be specifically affecting windows runs.

SwampDragons commented:
I'm looking into these test failures now.

SwampDragons commented:
Yeah, these windows failures were definitely not the fault of this builder. I'll merge a fix and then rebase again.

carlpett commented on builder/proxmox/config.go:
```diff
@@ -0,0 +1,196 @@
+	VMID   int    `mapstructure:"vm_id"`
+
+	Memory  int          `mapstructure:"memory"`
+	Cores   int          `mapstructure:"cores"`
```
:+1: for consistency. In this case there are both sockets and cores (per socket), though, does naming cores "cpus" make it a bit confusing? Setting sockets to 2 and cpus to 2 would result in a system with 4 "cpus", in some sense of the word.
I don't have a strong opinion here though, so if you've thought about the sockets+cores thing already I'll gladly change it :)

SwampDragons replied:
Ah, I see. Interesting. Yeah, leave it as cores then.

carlpett commented on builder/proxmox/config.go:
```diff
@@ -0,0 +1,196 @@
+	Type            string `mapstructure:"type"`
+	StoragePool     string `mapstructure:"storage_pool"`
+	StoragePoolType string `mapstructure:"storage_pool_type"`
+	Size            string `mapstructure:"size"`
```
Sure, it does go under a `disks` key already though, so it would stutter a bit? Eg
```json
      "disks": [
        {
          "disk_size": "5G",
          "..."
```

Consistency might be more valuable, though?

SwampDragons replied:
In this case, I think I prefer the consistency, even if it a bit of a stutter.

carlpett commented on builder/proxmox/config.go:
```diff
@@ -0,0 +1,196 @@
+
+	// Defaults
+	if c.ProxmoxURLRaw == "" {
+		c.ProxmoxURLRaw = os.Getenv("PROXMOX_URL")
```
That case is picked up in the "error checking block" at the end of the function. Although I apparently diverged a bit from that pattern with the boot key interval thing and did both default-setting and error checking at the same place :thinking: I'll see if I can restructure it a bit to be clearer.

SwampDragons replied:
This is probably fine then. I just read too fast and didn't catch the other block.

carlpett commented:
Thanks for fixing the tests! I spent some time yesterday trying (and failing) to understand how they could interact, glad that you found it :)

SwampDragons commented:
It was a fun one! There was some state getting left over by your boot command step test that was being picked up by a windows-shell test.

The reason that this has never happened for other boot command steps is.... no one wrote tests for them. You uncovered a bug in all of them, which is that none are cleaning up after themselves. I am going to write a PR this morning to fix the rest of them, but this one should be good to go apart from the couple of review comments I left. Thanks for your patience on this.

SwampDragons commented:
Okay, looked over your comments and I think the only thing I actually want you to touch is the disk_size name. Then we can merge. This is really good stuff.

carlpett commented:
Oh, tricky! :) I've pushed a change to `disk_size`.
Thanks for the reviewing and all the help!

SwampDragons commented:
Thanks! Updated the docs to match the code.

carlpett commented:
Ah, forgot that bit, of course... Thanks!
:tada:

rickard-von-essen commented:
Awesome! :beers:

azr commented:
Hello there @carlpett, I could not find to find the vendored go proxmox code from this PR online and this is preventing contributors from building packer using go modules.
Is it possible that you have changed the vendors only in the vendor folder or may be that the code is somewhere else ? Here's the issue: https://github.com/hashicorp/packer/issues/7483

to quote the error:
```
# github.com/hashicorp/packer/builder/proxmox
builder/proxmox/builder.go:112:15: c.GetVmAgentNetworkInterfaces undefined (type *proxmox.Client has no field or method GetVmAgentNetworkInterfaces)
builder/proxmox/step_convert_to_template.go:23:5: cannot use &proxmox.Client literal (type *proxmox.Client) as type templateConverter in assignment:
	*proxmox.Client does not implement templateConverter (missing CreateTemplate method)
builder/proxmox/step_start_vm.go:27:3: unknown field 'Agent' in struct literal of type proxmox.ConfigQemu
```

carlpett commented:
Uhoh. Yes. My PR to add it upstream didn't get any attention, and I forgot that. So sorry!
Perhaps the easiest way would be to point to my fork for now? 
I'll ping the maintainers in the PR as well

azr commented:
No worries, I think you can also point to a commit from your PR using:

```
GO111MODULE=on go get github.com/usr/proj@{commit}
```

carlpett commented:
If you could, that'd be great, thanks! I'm not on my dev machine right now. Here is the fork: https://github.com/carlpett/proxmox-api-go
The PR on the original repo doesn't contain all of it, since I wanted to be kind and split the changes over multiple PRs... If it is just for go modules, perhaps you could just do a `replace`?

azr commented:
There is still missing code in your PR https://github.com/Telmate/proxmox-api-go/pull/22/files for it to pass.

The best choice here is that you have your PR against their repo with everything needed, since it's their library; and we can have packer point to that PR. Using: 

`GO111MODULE=on go get github.com/Telmate/proxmox-api-go@$THE_LAST_COMMIT_OF_YOUR_PR`

carlpett commented:
Sure, I'll open another PR as soon as I'm back at my dev machine (an hour or so).
Sorry for the mess

azr commented:
No worries, I have to go out for today, please open a PR agains packer once you do !

carlpett commented:
@azr I got some nice help yesterday evening with getting everything actually merged into master in the library. I see that you reverted this in the meantime though - should I open a new PR with the builder again, or what is the best approach?

azr commented:
Hello @carlpett ! Awesome, thanks. I have #7490 open, I will update the vendor/deps an if everything is green, merge it.

azr commented:
Merged !

zaenk commented:
This plugin is is great! Thank you!

I can't wait for the release - so I built packer from master locally to play with this plugin, and I have some issues:

* The [docs](https://github.com/hashicorp/packer/pull/7391/files?file-filters%5B%5D=.md#diff-4d8ad6a99d84244b1840b1c869d3edd4R74) says, possible values of `os` might be `linux`, `windows`, `solaris` or `other`, but setting `linux` (for ProxMox 5.4-3) gives the following error

```
==> proxmox: Creating VM
==> proxmox: Error creating VM: 400 Parameter verification failed., error status: {"errors":{"ostype":"value 'linux' does not have a value in the enumeration 'other, wxp, w2k, w2k3, w2k8, wvista, win7, win8, win10, l24, l26, solaris'"},"data":null} (params: map[agent:1 cores:1 cpu:host description:Packer ephemeral build VM ide2:local:iso/ubuntu-18.04.2-server-amd64.iso,media=cdrom memory:512 name:base-packer net0:virtio=1A:C6:6E:4A:4B:D0,bridge=vmbr1 onboot:false ostype:linux scsi0:media=disk,size=5G,file=datassd:vm-199-disk-0 sockets:1 vmid:199])
```

Setting the option to `l26` fixed this issue.

Also the the VM disk image was not cleaned up after this failure.

* With a correct config the build always fails right after boot.

```
proxmox output will be in this color.

==> proxmox: Creating VM
==> proxmox: Starting VM
==> proxmox: Starting HTTP server on port 8986
==> proxmox: Waiting 10s for boot
==> proxmox: Stopping VM
==> proxmox: Deleting VM
Build 'proxmox' errored: unexpected EOF

==> Some builds didn't complete successfully and had errors:
--> proxmox: unexpected EOF

==> Builds finished but no artifacts were created.
```

The [crash log](https://gist.github.com/zaenk/790087fc2fd2d6a0c8d004efd282effa) gives away the culrpit https://github.com/hashicorp/packer/blob/master/builder/proxmox/step_type_boot_command.go#L69

`uint` should be used here like [other places in the code](https://github.com/hashicorp/packer/search?q=state.Get%28%22http_port%22%29&unscoped_q=state.Get%28%22http_port%22%29)

PS: If this is not the correct place for this kind of feedback, then please, point me to the right direction.

carlpett commented:
@zaenk Glad you like it :) I think creating an issue would be the best place in this situation, generally,
I'll create one and we can take it from there!
